partner_invoicing_mode/models/sale_order.py

```diff
@@ -1,10 +1,118 @@
+        for order in self:
+            order.one_invoice_per_order = order.partner_invoice_id.one_invoice_per_order
+
+    @api.model
+    def cron_generate_standard_invoices(self):
+        company_ids = self._get_companies_standard_invoicing()
+        if company_ids:
+            self.generate_invoices(company_ids)
+
+    @api.model
+    def _get_generate_invoices_state_domain(self):
+        return [("invoice_status", "=", "to invoice")]
+
+    @api.model
+    def _get_generate_invoices_domain(self, companies, invoicing_mode="standard"):
+        return AND(
+            [
+                self._get_generate_invoices_state_domain(),
+                [
+                    ("invoicing_mode", "=", invoicing_mode),
+                    ("company_id", "in", companies.ids),
+                ],
+            ]
+        )
+
+    @api.model
+    def generate_invoices(
+        self,
+        companies=None,
+        invoicing_mode="standard",
+        last_execution_field="invoicing_mode_standard_last_execution",
+    ):
+        """
+        Generate invoices in job queues depending on the invoicing
+        mode (stadndard by default)
+        """
+        if companies is None:
+            companies = self.env.company
+        domain = self._get_generate_invoices_domain(
+            companies=companies, invoicing_mode=invoicing_mode
+        )
+        saleorder_groups = self.read_group(
+            domain,
+            ["partner_invoice_id", "sale_ids:array_agg(id)"],
+            groupby=self._get_groupby_fields_for_invoicing(),
+            lazy=False,
+        )
+        for saleorder_group in saleorder_groups:
+            self.with_delay()._generate_invoices_by_partner(saleorder_group["sale_ids"])
+        companies.write({last_execution_field: Datetime.now()})
+        return saleorder_groups
+
+    @api.model
+    def _get_groupby_fields_for_invoicing(self):
+        """Returns the sale order fields used to group them into jobs."""
+        return ["partner_invoice_id", "payment_term_id"]
+
+    def _get_generated_invoices(self, partition):
+        return self._create_invoices(grouped=partition, final=True)
+
+    def _generate_invoices_by_partner(self, saleorder_ids):
+        """Generate invoices for a group of sale order belonging to a customer."""
+        sales = (
+            self.browse(saleorder_ids)
+            .exists()
+            .filtered_domain(self._get_generate_invoices_state_domain())
+        )
+        if not sales:
+            return "No sale order found to invoice ?"
+        # Create invoices using grouping when needed, so partition sales
+        invoice_ids = set()
+        for partition, sales in sales.partition(
+            lambda sale: sale.one_invoice_per_order
+        ).items():
+            invoices = sales._get_generated_invoices(partition=partition)
+            # Update each partner next invoice date
+            sales.partner_invoice_id._update_next_invoice_date()
```

Comment:
move this outside loop ?

Reply:
you may have concurrent update on this call as you have one job per `_get_groupby_fields_for_invoicing` and this can concern the same commercial entity

Reply:
In that case, the job is automatically postponed

Reply:
See : https://github.com/OCA/queue/blob/16.0/queue_job/controllers/main.py#L127

Reply:
but still worth moving the `_update_next_invoice_date()` outside the loop as it is independent from the partition and the the `_get_generated_invoices`

Reply:
@rousseldenis what about moving `_update_next_invoice_date` outside partition to prevent useless jobs?